### PR TITLE
syndicate: a simple implementation of P.O.S.S.E.

### PR DESCRIPTION
**Pull request by dabrady**
To start with, I've written a single silo adapter for [DEV.to](https://dev.to), because that's where I posted my first blog post and I'm biased.

### Prereqs and setup
Create a dummy repo and copy the provided `example.yml` file to its `.github/workflows` folder.

Create an account on [DEV.to](https://dev.to) and generate an API key for yourself in your account settings. Add that key as a secret to your dummy repo under the name `DEV_API_KEY`.

### Testing draft creation
Now create a fake Markdown post with a `.md` extension in your dummy repo located in a `pages/posts` directory, ensuring you define a minimal YAML frontmatter containing a `title` attribute, like so:

```markdown
---
title: A lovely post.
---

My, what a lovely post this is. Don't you just enjoy reading it?

Like and subscribe!
```

Commit, push to `master`, and check the Actions tab of your repo on Github: you should see the example workflow running, and can click through to see the details.

Expect the workflow to succeed with a few warnings about `Medium` and `Planet_Mars` being unsupported platforms.

Expect a commit to have been pooshed to `master` on behalf of `github-actions` which adds a `dev_silo_id` to the YAML frontmatter of your post.

Expect to see a new, unpublished draft in your DEV.to dashboard containing the contents of your dummy post.

### Testing post updates
Now make a change on Github to the dummy post you just created and poosh that change to `master`.

Go back to the Actions tab of your Github repository and expect to see a new workflow running that subsequently succeeds.

Navigate to your DEV.to dashboard and go to your syndicated draft: expect to see the changes you made on Github mirrored to your draft.

### CODE REVIEW

**Comment by dabrady** [comment]
@rhymiz Care to lend your Pythonic expertise in the form of sagely advice and comments? 😄🙏 This is my first Python project, so I'm mostly looking for feedback on project structure, best practices, and the overall levels of Pythonliness you perceieve.

I recommend not reviewing commit by commit, if you decide to help out: there are only 14 files with things in them, but there are exactly 100 commits. 😅

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,17 @@
+FROM python:3-alpine
+
+WORKDIR /action
+
+# Copy action metadata
+COPY LICENSE README.md requirements.txt ./
```
Comment (rhymiz) [inline Dockerfile]: don't think you need LICENSE and README.md in the docker container

File: syndicate/__init__.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        if silo in api_keys
+    }
+    if results:
+        return results
+    else:
+        action_warn("Sorry, can't do anything with that!")
+        return None
+
+### privates ###
+_API_KEY = lambda s: f"{s.upper()}_API_KEY"
```
Comment (rhymiz) [inline syndicate/__init__.py]: better to extract this into a function rather than using a lambda

File: syndicate/silos/dev.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+    This tries to create an **unpublished** draft. However, the 'published'
+    status can be overridden in the frontmatter of the post itself for a
+    "just do it" approach.
+
+    @see https://docs.dev.to/api/#operation/createArticle
+    """
+    assert api_key, "missing API key"
+    assert post, "missing post"
+    assert fronted(post).get('title'), "article is missing a title"
```
Comment (rhymiz) [inline syndicate/silos/dev.py]: as a reminder, `assert` will be ignored when Python is running with optimizations. assertions are meant for environments where Python is running in debug/dev mode.

File: syndicate/utils.py [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+
+# Memoize authentication and repo fetching.
+@functools.lru_cache(maxsize=1)
+def repo():
+    """
+    (MEMOIZED) Returns an authenticated reference to a repository object for the
+    repository this Github action is running in.
+    @see https://pygithub.readthedocs.io/en/latest/github_objects/Repository.html#github.Repository.Repository
+    """
+    assert os.getenv("GITHUB_TOKEN"), "missing GITHUB_TOKEN"
```
Comment (rhymiz) [inline syndicate/utils.py]: same comment about assertions

File: syndicate/utils.py [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+
+def fronted(post):
+    """
+    Returns the :py:class:`frontmatter.Post` representation of the given
+    :func:`~syndicate.utils.file_contents` object.
+
+    If `post` is actually already a `frontmatter.Post`, this is a no-op.
+    """
+    assert post, "missing post"
+    if type(post) == frontmatter.Post:
```
Comment (rhymiz) [inline syndicate/utils.py]: use `isinstance(object, other_object)` instead of `type(object) == other_object`

File: syndicate/utils.py [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+    change, and so we don't try to change them.
+
+    Returns a dictionary which is the response of the commit request.
+    """
+    assert silo_ids_by_path, "missing silo IDs"
+    assert fronted_posts_by_path, "missing fronted posts"
+
+    updated_fronted_posts_by_path = {}
+    silos_included = set()
+    for (path, silo_ids_by_silo) in silo_ids_by_path.items():
```
Comment (rhymiz) [inline syndicate/utils.py]: parens are optional here, I'd use the following instead:
```python
for path, silo_ids_by_silo in silo_ids_by_path.items():
    ...
```

File: syndicate/utils.py [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+        fronted_post = fronted_posts_by_path[path]
+
+        # Format:
+        # {
+        #     'dev_silo_id': 42,
+        #     'medium_silo_id': 'abc123',
+        #     ...
+        # }
+        new_silo_ids = {}
+        for (silo, sid) in silo_ids_by_silo.items():
```
Comment (rhymiz) [inline syndicate/utils.py]: same about the parens

File: entrypoint.py [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+if action_inputs['mark_as_syndicated']:
+    action_log("Marking newly syndicated posts...")
+    ## NOTE
+    # If silos were provided, commit only the results of this step. In the case
+    # where no silos were provided, commit all job results so far.
+    #
+    # This allows us to bundle syndications into as few or many commits as we
+    # want in our workflows.
+    ##
+    if not action_inputs['silos']:
```
Comment (rhymiz) [inline entrypoint.py:49]: will this key always exist? if not, you may want to use `action_inputs.get('silos')` which will return `None` when the key is not present, else you'll get a `KeyError`

File: Dockerfile (hunk 1 above)
Comment (dabrady) [reply Dockerfile]: I think you're right. Not sure how this evolved, probably came from an example I was following at some point lol. Will remove 👍

File: syndicate/silos/dev.py (hunk 3 above)
Comment (dabrady) [reply syndicate/silos/dev.py]: Good point 💯 Better to replace most/all of these `asserts` with actual input handling and raise value errors when they fail.

File: entrypoint.py (hunk 8 above)
Comment (dabrady) [reply entrypoint.py:49]: Good point. 👍 I'll confirm and update if necessary.

Comment (dabrady) [reply entrypoint.py:49]: The key will always exist. I'm setting it explicitly up here:

https://github.com/dabrady/syndicate/blob/edc91a3a485e81f789f183b7cee96f29a571c1fd/entrypoint.py#L9-L12

File: syndicate/silos/dev.py (hunk 3 above)
Comment (dabrady) [reply syndicate/silos/dev.py]: Done in https://github.com/dabrady/syndicate/pull/1/commits/a7606e890570c849e77faa35c14f33c382c8d04b

File: syndicate/__init__.py (hunk 2 above)
Comment (dabrady) [reply syndicate/__init__.py]: Done in https://github.com/dabrady/syndicate/pull/1/commits/48f84920550bd3e6e5d1fd3e25efea0b939d00ac

File: Dockerfile (hunk 1 above)
Comment (dabrady) [reply Dockerfile]: Done in https://github.com/dabrady/syndicate/pull/1/commits/54c83875e4607667c7b90877617f195e87c2916f

File: syndicate/utils.py (hunk 5 above)
Comment (dabrady) [reply syndicate/utils.py]: Done in https://github.com/dabrady/syndicate/pull/1/commits/9ade4e12ad43c16807f6ab4517de4d7c90ba9c6d

File: syndicate/utils.py (hunk 6 above)
Comment (dabrady) [reply syndicate/utils.py]: Done in https://github.com/dabrady/syndicate/pull/1/commits/598380b3ecc84072f3dbb1eda6d9924a51d671c7

**Comment by dabrady** [comment]
@rhymiz Feedback addressed! Please re-review 🙏

**Comment by dabrady** [comment]
Changes verified on a test branch with this commit: https://github.com/dabrady/syndicate/commit/c2da546cf495ea23f26a925894330ed76bca1e3b

Job output and resulting draft on DEV.to:
image

Upstream commit generated to test branch: https://github.com/dabrady/syndicate/commit/dbbefd75ec26dcc53ee9fa812f99acca06f25c50

**Review by rhymiz — APPROVED** [review]
gucci!

**Comment by dabrady** [comment]
Thanks for your eyes and brains, @rhymiz! 🙏 😄

**Comment by dabrady** [comment]
And it's live: https://github.com/marketplace/actions/syndicate-elsewhere

**Pull request merged**